tfx/extensions/google_cloud_kubernetes/pusher/executor.py

```diff
@@ -0,0 +1,185 @@
+
+# Keys for custom_config.
+_CUSTOM_CONFIG_KEY = 'custom_config'
+
+# Number of serving replicas.
+NUM_REPLICAS_KEY = 'num_replicas'
+
+# Name of serving model.
+MODEL_NAME_KEY = 'model_name'
+
+# Model uri environment variable key.
+_MODEL_URI_ENV_KEY = 'MODEL_URI'
+
+# Model name environment variable key.
+_MODEL_NAME_ENV_KEY = 'MODEL_NAME'
+
+# Model base path environment variable key.
+_MODEL_BASE_PATH_ENV_KEY = 'MODEL_BASE_PATH'
+
+class Executor(tfx_pusher_executor.Executor):
+  """Deploy a model to GKE cluster with Tensorflow Serving."""
+
+  def DeployTFServingService(self) -> None:
+    """Creates the model serving service with TF Serving."""
+    client_api = kube_utils.make_core_v1_api()
+    with open(path.join(path.dirname(__file__),
+                        "yaml", "serving-service.yaml")) as f:
+      svc = yaml.safe_load(f)
+      try:
+        resp = client_api.create_namespaced_service(
+            body=svc, namespace="default")
+        logging.info("Model Service created. status='%s'" % str(resp.status))
+      except rest.ApiException:
+        # Since the model service yaml is static, no update is needed.
+        logging.info("Model Service unchanged.")
+
+  def DeployTFServingDeployment(
+      self,
+      model_name: Text,
+      model_uri: Text,
+      num_replicas: int,
+  ) -> None:
+    """Creates or updates the model serving deployment with TF Serving.
+
+    Args:
+      model_name: Name of the model being served, used as part of the local
+        model path in the serving container as well as the exposed api endpoint
+        to the client.
+      model_uri: Uri of the serving model output from which the container
+        will download the model.
+      num_replicas: Number of serving replicas.
+  """
+    client_api = kube_utils.make_externsions_v1_beta1_api()
+    with open(path.join(path.dirname(__file__),
+                        "yaml",
+                        "serving-deployment.yaml")) as f:
+      dep = yaml.safe_load(f)
+      env_vars = [
+          {'name': _MODEL_BASE_PATH_ENV_KEY, 'value': '/models'},
+          {'name': _MODEL_URI_ENV_KEY, 'value': model_uri},
+          {'name': _MODEL_NAME_ENV_KEY, 'value': model_name},
+      ]
+      spec = dep['spec']
+      # Configure the number of replicas.
+      spec['replicas'] = num_replicas
+      # Configure the environment variables.
+      spec['template']['spec']['containers'][0]['env'] = env_vars
+      try:
+        resp = client_api.create_namespaced_deployment(
+            body=dep, namespace="default")
+        logging.info("Deployment created. status='%s'" % str(resp.status))
+      except rest.ApiException:
+        logging.info(dep)
+        resp = client_api.patch_namespaced_deployment(
+            name=dep['metadata']['name'],
+            body=dep, namespace="default")
+        logging.info("Deployment updated. status='%s'" % str(resp.status))
+
+  def Do(self, input_dict: Dict[Text, List[types.Artifact]],
```

Comment:
I know you wanted to revamp this, so I will hold off on reviewing this until you finish that refactor.

Reply:
Thanks, refactoring is finished.